### DISCUSSION TITLE
Using Redis Graph for native graph support

### DISCUSSION THREAD

**Discussion by pokonski** [chaps-io/gush#95] [Ideas]
Right now Gush implements DAGs internally on the Ruby side which proved to be problematic at scale due to race conditions when multiple jobs can trigger the same downstream job (which we put a band-aid over in the form of mutexes, those in turn slow down workflows with thousands of jobs)

For the longest time I have been considering using https://redis.io/docs/stack/graph/ for storage and querying jobs, but it was problematic to include in Redis instances. 

But I see they greatly simplified the setup, so asking users to run a RedisGraph instance is no longer hard (Docker image is trivial to run and I assume many people already use docker).

The downside is that people using ElastiCache or other paywalled/managed Redis solutions will not have an option to run easily. But I still think it would simplify Gush greatly and possibly improve performance/reliability.

Any thoughts?
cc @joshRpowell @thukim @Saicheg

**Comment by pokonski** [comment]
We could possibly add support for multiple adapters, the original and RedisGraph. So we don't alienate existing users.

**Comment by machty** [comment]
FYI RedisGraph end-of-life was announced July this year: https://redis.com/blog/redisgraph-eol/?_gl=1*18ggq17*_ga*NzEwOTI1ODY4LjE2OTY5MTY2MDc.*_ga_8BKGRQKRPV*MTY5NjkxNjYwNy4xLjAuMTY5NjkxNjYwNy42MC4wLjA.&_ga=2.145394356.1283594008.1696916608-710925868.1696916607
